File: addBinary.js
/**
 * @param {string} a
 * @param {string} b
 * @return {string}
 */
const addBinary = function (a, b) {
    let i = a.length - 1;
    let j = b.length - 1;
    let carry = 0;
    let result = "";

    while (i >= 0 || j >= 0) {
        let sum = carry;

        if (i >= 0) {
            sum += parseInt(a[i]);
            i--;
        }

        if (j >= 0) {
            sum += parseInt(b[j]);
            j--;
        }

        if (sum === 0 || sum === 1) {
            result += sum.toString();
            carry = 0;
        } else if (sum === 2) {
            result += "0";
            carry = 1;
        } else {
            result += "1";
            carry = 1;
        }
    }

    if (carry > 0) {
        result += "1";
    }

    return result.split("").reverse().join("");
};

console.log(addBinary('11', '1'))

File: leetcode_rust/src/problems/add_binary.rs
pub struct Solution;

impl Solution {
    pub fn add_binary(a: String, b: String) -> String {
        let mut result = String::new();
        let mut carry = 0;
        let mut al = a.len();
        let mut bl = b.len();

        while al > 0 || bl > 0 {
            let mut sum = carry;
            if al > 0 {
                sum += a.chars().nth(al - 1).unwrap() as i32 - '0' as i32;
                al -= 1;
            }
            if bl > 0 {
                sum += b.chars().nth(bl - 1).unwrap() as i32 - '0' as i32;
                bl -= 1;
            }
            result.push((sum % 2 + '0' as i32) as u8 as char);
            carry = sum / 2;
        }

        if carry > 0 {
            result.push('1');
        }

        result.chars().rev().collect()
    }
}
